NODE-1209: clients handle all CLValues for query and deploy arguments

birchmd opened the pull request:
### Overview
This PR exposes the full `CLValue` enum all the way out to the clients. Now from the CLI you can query any value type, as well as give any value type as an argument for a deploy.

### Which JIRA ticket does this PR relate to?
https://casperlabs.atlassian.net/browse/NODE-1209

### Complete this checklist before you submit this PR
- [ ] This PR contains no more than 200 lines of code, excluding test code.
- [x] This PR meets [CasperLabs coding standards](https://casperlabs.atlassian.net/wiki/spaces/EN/pages/16842753/Coding+Standards).
- [x] If this PR adds a new feature, it includes tests related to this feature.
- [x] You assigned one person to review this PR.
- [x] Your GitHub account is linked with our [Drone CI](https://drone-auto.casperlabs.io/) system. This is necessary to run tests on this PR.
- [x] Do not forget to run `bors r+` if GitHub policy is not enforced, e.g. when merging into another feature branch. It may be omitted under some circumstances if this PR intentionally assumes that integration tests will fail but will be fixed with the future PRs.

### Notes
There are lots of TODOs about stack safety. I will go in and fix those, but I wanted to get the PR out since it is really big and touches lots of different parts of the code base.

I also apologize the commits are not as granular as usual. I took a bit of a meandering path to get here, so the original commit history is pretty dirty. I squashed a bunch of things down into a single big commit once it got to a state where everything seemed to work.

birchmd commented:
bors try

aakoshh commented:
Dude :drooling_face:

birchmd commented:
bors retry

aakoshh commented:
I experience severe stateCLTypeStoredValueInstanceToProtoConverter blindness, I'll have to digest this slowly; but man, thank you for putting in all the effort!

aakoshh commented on protobuf/io/casperlabs/casper/consensus/consensus.proto:
```diff
@@ -89,14 +89,20 @@ message Deploy {
     }
 
     message Arg {
+        reserved 2; // old state.Value representation of args
```
Why not add a new field to the `Code` and keep it really backwards compatible? Done like this, an old client will appear not to have sent any arguments at all and the deploy hash will be invalid.

Doing it in the following way would make sure the node can process data coming from any old client (provided the fallback conversion is also moved from the Scala client to the node).
```
message Code {
    ...

        repeated LegacyArg legacy_args = 6;

        // Keyword arguments usng CLValue. Note that while the execution engine works with the original
        // positional argument passing style, the order of the arguments matters.
        repeated Arg args = 7;
}
```

birchmd replied:
Do we need to support old clients? I thought an advantage to getting this in before the March release would be that we don't really have users so we can make breaking API changes as long as we update all the clients we ship.

The main reason I left the old JSON format around was because I didn't want to try to find and update all the bash scripts we have lying around (and also because of the verbosity problem you mentioned, which in the future will hopefully be fixed).

aakoshh replied:
I understand. I don't know if we have to. @TomVasile needs to be aware that once this PR is in, Clarity, Python client, Scala client, everything has to be on the newest version at the same time.

TomVasile replied:
bummer, that means we cant update devnet\testnet's Clarity without a release?

aakoshh replied:
Yes @TomVasile, this PR, as it is, will break any existing client that either
* wants to send deploy arguments 
* want to read query results (e.g. run balance queries)

And any client that would be updated without the node being updated at the same time will also not work as they will be expecting to be able to send new style arguments (which are ignored) and look for new style query results (not present).

We can hold on to it until you're ready.

TomVasile replied:
just thinking out loud of what we will need to do.

we use an older version of the python client in the lrt's due to a bug. We will need to move off of that version. 

clarity also will need to be tied into the role. I think @piotr-putit had worked on this role already so this is likely okay.

piokuc replied:
@TomVasile if I remember correctly the problem was caused by a conflict of a crypto library version that we started to use in the Python client, that is also used by Ansible. The solution was to install our client and scripts in a Python virtual environment such as pipenv, rather than globally.

TomVasile replied:
yea that sounds right. We havent had a need to fix it until now. I'll take a look later today.

aakoshh commented on protobuf/io/casperlabs/node/api/casper.proto:
```diff
@@ -155,7 +155,8 @@ message BatchGetBlockStateRequest {
 
 message BatchGetBlockStateResponse {
-    repeated io.casperlabs.casper.consensus.state.Value values = 1;
+    reserved 1; // previously repeated io.casperlabs.casper.consensus.state.Value
```
_If_ you wanted to leave a deprecation period this could be served along with the the new field.

aakoshh commented:
I also found the way protobuf can be constructed a bit too confusing, too many ways. Would be good to have 1 convenient way of doing it. I'm also not sure that using both `cltype.CLValueInstance` and `state.CLValueInstance` alternatingly to make requests is not more confusing than if the user only had access to the proto version, as in Python and TypeScript.

birchmd commented on build.sbt:
```diff
@@ -514,7 +514,7 @@ lazy val client = (project in file("client"))
     )
   )
-  .dependsOn(models, graphz)
+  .dependsOn(models, graphz, smartContracts)
```
Yeah, good point. I was originally putting everything in `smartContracts` because that is only where is was getting used, but the type definitions probably belong in `models` and then this dependency would not be needed.

aakoshh replied:
Putting them into `models` sounds like a good idea. At least the `client` wouldn't get all the extra classes from `ipc.proto`.

birchmd replied:
Moved to `models` in https://github.com/CasperLabs/CasperLabs/pull/1687/commits/1feb9cded4c93db73c07836e284507bdc179a642

birchmd commented on client/src/test/scala/io/casperlabs/client/configuration/ArgsSpec.scala:
```diff
@@ -1,18 +1,352 @@
+      {"name" : "i64", "value" : {"cl_type" : { "simple_type" : "I64" }, "value" : { "i64" : $long }}},
+
+      {"name" : "u8", "value" : {"cl_type" : { "simple_type" : "U8" }, "value" : { "u8" : $byte }}},
+
```
Yes, I agree it is too verbose. I would like the type definition to be optional for all basic types and some non-basic types too (like non-empty options and lists). I'm not sure what the best way to go about doing that is though and I figured it could be done as a future enhancement.

aakoshh replied:
We just discussed with @zie1ony that he'd like to go towards a DSL, for example:
```bash
casperlabs-client deploy --session-args "$(casperlabs-args name1:u8:3, name2:string:"hello world", my_tuple:tuple2:(u64:123, i32:123))" ...
```

He mentioned this should also work in the browser, so I suggested we could have it written in the TypeScript SDK we are publishing as part of Clarity, and run it via `node`, like [here](https://github.com/CasperLabs/CasperLabs/tree/dev/explorer#fund-the-faucet), but packaged as an executable.

It wouldn't have to support everything, only what's convenient, but it might be a way out of the JSON conundrum. I don't know if someone sending a complicated data structure will be able to type out the `cl_type` for it. Having it inferred either on the client or the server side would make it easier for them to avoid little mistakes.

piokuc replied:
When we introduced ability to pass contract arguments on command line in a JSON protobuf format it was really meant to be a temporary measure. Nobody liked that, I think we all felt something more concise and elegant would be possible but we left developing it for later. Now, with this PR the JSON representation of args is getting even more verbose so I think it is high time to make some decisions re DSL for specifying values of arguments.

Once we have the contract headers and so information about type signatures of contracts the user will only need to pass arguments' values on command line, not their types. Client will use the info on types to a) construct appropriate protobuf objects, b) check types of values passed by the user match the types declared in contract header. So it could look like this:

```
casperlabs-client deploy --session-args-values '3,"hello world",(123, 123)'
```

I do remember we have had plans for named parameters but it seems we are not going to have them before the March release so, in my opinion, we should revisit the plans and say we would support named parameters in the future, but as an addition to positional ones, similarly to popular programming languages like, e.g. Python.

Now, obviously we don't have the contract headers in place yet and so the clients don't have the contract arguments' types information, therefore the user has to pass this info to the clients. I think it will be best to pass this information separately:

```
casperlabs-client deploy  --sesions-args-types "u8,string,tuple[u64,i32]" \
                          --session-args-values '3,"hello world",(123, 123)'
```

After we implement contract headers we can make `-sesions-args-types` optional.

birchmd replied:
I really like the idea of a custom DSL. This definitely seems much better then the JSON version. I'm not sure whether it is better to give the type separately or with the value (as Maciej suggested). Probably we should put some thought into this and write a design doc we come to agreement on before doing an implementation. I've created an ecosystem ticket for this:  https://casperlabs.atlassian.net/browse/ECO-343

zie1ony replied:
Sure, I'll prepare the design, we'll all review it and then we can implement it.

birchmd commented on client/src/test/scala/io/casperlabs/client/configuration/ArgsSpec.scala:
```diff
@@ -1,18 +1,352 @@
+      state.CLValueInstance(
+        clType = ProtoMappings.toProto(CLType.Bool).some,
+        value = ProtoConstructor.bool(bool).some
+      )
```
You're right, I was a little lazy and inconsistent with my construction of the protobuf instances. In general I don't like the protobuf classes because they are much more verbose. I think the scala ADT is pretty nice though, so much prefer using it. Of course, at the rpc boundaries we need to convert to protobuf, but maybe the standard I should follow is to use the scala ADT everywhere and only convert to protobuf as needed.

aakoshh replied:
Agreed the classes generated by scalapb are mostly only usable with some DSL support to hide the optionality and figure out types. I thought your original `CLTyped` approach would be excellent here.

birchmd replied:
I added a DSL in https://github.com/CasperLabs/CasperLabs/pull/1687/commits/211ef6e02ed1048336ef3141046783416c0be1e0
I think it makes things cleaner and more consistent when needing to create protobuf instances

zie1ony commented:
@aakoshh Thanks for pointing me this PR. I remember that the json format was design with the idea that people will not have to compose those by hand. The truth is that I had to write it by hand so many times. 
I suggest to have simpler, more user friendly format, eg:
```
name1:u8:3, name2:string:"hello world", my_tuple:tuple2:(t1:u64:123, t2:t2:i32:123)
```
Hopefully it's self-explanatory.

piokuc commented on integration-testing/test/test_args_R1.py:
```diff
@@ -57,8 +57,9 @@ def test_args_to_json():
         ABI.args(args), ensure_ascii=True, sort_keys=True, indent=2
     )
-    assert json_str1 == json_str2
+    # TODO: fix this test to use new json format
```
I can do that later 👍 Better, yet, if we come up with a better notation of specifying types and values, as I mentioned in the other comment, and stop using the JSON protobuf representation altogether!

piokuc commented on protobuf/io/casperlabs/casper/consensus/state.proto:
```diff
@@ -79,6 +79,92 @@ message CLValue {
+    message Option {
+        Value value = 1;
+    }
+    // Technically these types can represent hetrogeneous lists and maps,
```
Typo: "hetrogeneous" => "heterogeneous"

birchmd replied:
Done in https://github.com/CasperLabs/CasperLabs/pull/1687/commits/80d2b5a68756eab0230e0611fef9e9ce1a2a438d

piokuc commented on protobuf/io/casperlabs/casper/consensus/state.proto:
```diff
@@ -79,6 +79,92 @@ message CLValue {
+            Map map_value = 18;
+            Tuple1 tuple1_value = 19;
+            Tuple2 tuple2_value = 20;
+            Tuple3 tuple3_value = 21;
```
What is the benefit of having these three types of `Tuple` instead of using `FixedList` consistently?

piokuc replied:
OK, I guess the idea is that `FixedList` is homogeneous but `Tuple{1,2,3}` possibly heterogeneous. Maybe `Array` would be a better name for `FixedList` then? I suppose it can be slightly problematic due to common expectation that arrays should be multidimensional, but I can imagine that due to the recursive definition of `Value` we can have multidimensional arrays encoded as array of arrays, can't we?

Having a general / arbitrary length `Tuple` would be great. I gather from your comments on line 129 that there are some issues with implementing it. I'm curious what it is.

birchmd replied:
Yes, that's right the tuple data types allow for heterogeneous pairs and triples (e.g. `(i32, String)`).

`Array` seemed like too much of an overloaded term use, but you are right that in Rust these collections with a statically known length are indeed called arrays and you could certainly encode multi-dimensional such collections as arrays of arrays. `FixedList` was meant to be short for `FixedLengthList`, and I think it is important to emphasize this being the distinguishing feature of this collection over the regular `List` (as opposed to `Array` where people may mistake the difference as being performance related).

I would also like to see arbitrary sized tuples, but it is difficult to write generic code for them. Generally, you need to resort to doing some meta-programming (often in the form of macros) to generalize over the arity of the tuple. For now, on the rust side we only support up to 3-tuples for this reason, and I kept that consistent all the way out to the client level.

piokuc commented on protobuf/io/casperlabs/casper/consensus/state.proto:
```diff
@@ -79,6 +79,92 @@ message CLValue {
+    // however it will be considered illegal to do so because there is no way to
+    // assign a CLType to such an object.
+    message List {
+        repeated Value values = 1;
```
So now we can have (homogeneous) lists of lists, right? Nice!

birchmd commented on client/src/main/scala/io/casperlabs/client/configuration/Args.scala:
```diff
@@ -57,11 +59,157 @@ object Args {
+
+  private def legacyArgToArg(legacy: LegacyArg): Arg =
+    Arg(name = legacy.name, value = legacy.value.map(legacyValueToInstance))
+
```
Someday we hopefully won't need this logic in either place. But I suppose the same argument could be made about parsing the eventual DSL we come up with (unless it is always passed through some other tool as I think Akosh was suggesting).

birchmd replied:
The main reason I did not do it server-side was because it seemed like it would involve specifying the rpc in a way which still included the old args types, which I did not want to do. I hope the rpc definition will be reasonably stable after this PR.

goral09 commented on client/src/main/scala/io/casperlabs/client/DeployRuntime.scala:
```diff
@@ -84,33 +87,30 @@ object DeployRuntime {
 
   private def longArg(name: String, value: Long) =
-    arg(name, Deploy.Arg.Value.Value.LongValue(value))
+    arg(name, cltype.ProtoMappings.toProto(cltype.CLValueInstance.U64(value)))
```
The life-span of Scala ADT encoded arguments is very short.

goral09 commented on smart-contracts/src/main/scala/io/casperlabs/smartcontracts/cltype/CLValueInstance.scala:
```diff
@@ -5,12 +5,56 @@ import cats.implicits._
+    case CLValueInstance.List(values, _) =>
+      ToBytes.toBytes(values.size) ++ values.toIndexedSeq.flatMap(_.valueBytes)
+
+    case CLValueInstance.FixedList(values, _, _) => values.toIndexedSeq.flatMap(_.valueBytes)
```
it doesn't need to encode its size?

birchmd replied:
No, the size is not encoded in the value bytes because it is encoded in the type

goral09 replied:
Thanks. I am already lost here.

birchmd commented:
@aakoshh I think this PR is good to go now. I added a DSL for the protobuf creation in the scala code as you suggested, and moved the definitions of things into `models`. Let me know if you think it needs anything else. I'll also wait for @TomVasile 's approval before merging.

birchmd commented:
bors try

TomVasile commented:
@birchmd one clarification on this: so does this mean going forward the python client always has to be from the same git hash as the node (and etc)? or is it that after this merges only previous released version wont work on dev?

The reason I ask is this could be a bit of a pain in the stests framework that uses the python client. We have lrt1 and lrt2 often running different git hashes.

birchmd commented:
bors retry

TomVasile commented:
@birchmd converted our role to use the python client from dev so we should be set

aakoshh commented on build.sbt:
```diff
@@ -185,7 +185,9 @@ lazy val models = (project in file("models"))
       catsCore,
+	  catsFree,
       magnolia,
+	  refinement,
```
Strange indent.

birchmd replied:
Fixed in https://github.com/CasperLabs/CasperLabs/pull/1687/commits/b1eeaebab882a0489890782162ab5f56560ee1b9

aakoshh commented on models/src/main/scala/io/casperlabs/models/cltype/ByteArray32.scala:
```diff
@@ -15,6 +15,13 @@ object ByteArray32 {
+      xs.iterator.zip(ys.iterator).exists {
+        case (x, y) => x < y
+      }
+  }
```
I might be wrong, but it looks like this will say `a` is less than `b` if _any_ byte in `a` is less than the one in `b` in the same position, so for example `lt([2,1], [0,0])` would be `true`? Shouldn't it loop until it finds the first position where they are not equal and return what it finds there?

birchmd replied:
Yes, you are totally right. I'll fix it.

birchmd replied:
Fixed in https://github.com/CasperLabs/CasperLabs/pull/1687/commits/9cfe2aaed8000afd24e608e8c0d55981432e303c

aakoshh commented on models/src/main/scala/io/casperlabs/models/cltype/CLValueInstance.scala:
```diff
@@ -266,6 +271,33 @@ object CLValueInstance {
 
     case class InvalidLength(valueLength: Int, typeLength: Int) extends Error
+
+    case object UnorderedElements extends Error
```
What's the use case for this error?

birchmd replied:
When serializing a `Map` the elements must have a well-defined order on them so the serialization is deterministic. This error happens when the elements cannot be ordered (because one is not defined).

aakoshh replied:
Thanks; maybe worth adding a comment about it.

birchmd replied:
Comment added in https://github.com/CasperLabs/CasperLabs/pull/1687/commits/eec51de5dcb154bb316107575b4ff93538465d93

aakoshh commented on models/src/main/scala/io/casperlabs/models/cltype/CLValueInstance.scala:
```diff
@@ -266,6 +271,33 @@ object CLValueInstance {
+      ByteArray32.lt(seed1, seed2) || (seed1 == seed2 && ByteArray32.lt(hash1, hash2))
+
+    // TODO: complete ordering implementation
+    case _ => throw new Exception("Ordering not implemented for recursive CLValueInstances")
```
Are you likely to find mismatched instances, i.e. trying to compare `U64` to `String`?

birchmd replied:
I don't think so. This is really only used to order they keys in a `Map`, and they will all need to be the same `CLType` if it is a valid `Map` instance.

aakoshh replied:
Ah, okay, I had no idea where ordering was used, thanks.

birchmd commented:
> Looks like manual mappings still have the tendency to turn into a full time job.

Hopefully manual mappings will never be required again since everyone will be able to use the `toProto` and `fromProto` functions for any conversions.

birchmd commented:
bors try

aakoshh commented on build.sbt:
```diff
@@ -185,7 +185,9 @@ lazy val models = (project in file("models"))
       catsCore,
+  	  catsFree,
       magnolia,
+	    refinement,
```
There are tabs here. Just sayin'

birchmd replied:
🤦‍♂ I didn't notice in my editor, thanks for pointing it out.

Fixed in https://github.com/CasperLabs/CasperLabs/pull/1687/commits/a8e3683b23085e4dabd7a35f5fb8d230a456f4b0

aakoshh commented on integration-testing/test/test_single_network_one_R1.py:
```diff
@@ -1017,7 +1017,7 @@ def test_invalid_bigint(one_node_network):
     # Use a malformed BigInt contract argument
     node = one_node_network.docker_nodes[0]
-    cli = DockerCLI(node)
+    cli = CLI(node, "casperlabs_client")
```
What's the need for this change?
cc @piokuc

birchmd replied:
It's described in the commit message https://github.com/CasperLabs/CasperLabs/pull/1687/commits/23a1f0abb4bb50ece77e1b6bd29f46c747c847c2

Essentially it is because the scala client now rejects the bad input immediately, but the python client still lets it through to the node, and allows the test to proceed as before.

birchmd commented:
bors r+
